File: Team00000/src/main/java/org/firstinspires/ftc/team00000/tuning/TuningOpModes.java
package org.firstinspires.ftc.team00000.tuning;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.config.reflection.ReflectionConfig;
import com.acmerobotics.roadrunner.MotorFeedforward;
import com.acmerobotics.roadrunner.Pose2d;
import com.acmerobotics.roadrunner.ftc.*;
import com.qualcomm.hardware.lynx.LynxModule;
import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.eventloop.opmode.OpModeManager;
import com.qualcomm.robotcore.eventloop.opmode.OpModeRegistrar;
import org.firstinspires.ftc.robotcore.internal.opmode.OpModeMeta;
import org.firstinspires.ftc.team00000.roadrunner.MecanumDrive;
import org.firstinspires.ftc.team00000.roadrunner.ThreeDeadWheelLocalizer;
import org.firstinspires.ftc.team00000.roadrunner.TwoDeadWheelLocalizer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class TuningOpModes {
    public static final Class<?> DRIVE_CLASS = MecanumDrive.class;

    public static final String GROUP = "quickstart";
    public static final boolean DISABLED = false;

    private TuningOpModes() {}

    private static OpModeMeta metaForClass(Class<? extends OpMode> cls) {
        return new OpModeMeta.Builder()
                .setName(cls.getSimpleName())
                .setGroup(GROUP)
                .setFlavor(OpModeMeta.Flavor.TELEOP)
                .build();
    }

    @OpModeRegistrar
    public static void register(OpModeManager manager) {
        if (DISABLED) return;

        DriveViewFactory dvf;
        if (DRIVE_CLASS.equals(MecanumDrive.class)) {
            dvf = hardwareMap -> {
                MecanumDrive md = new MecanumDrive(hardwareMap, new Pose2d(0, 0, 0));

                List<Encoder> leftEncs = new ArrayList<>(), rightEncs = new ArrayList<>();
                List<Encoder> parEncs = new ArrayList<>(), perpEncs = new ArrayList<>();
                if (md.localizer instanceof MecanumDrive.DriveLocalizer) {
                    MecanumDrive.DriveLocalizer dl = (MecanumDrive.DriveLocalizer) md.localizer;
                    leftEncs.add(dl.leftFront);
                    leftEncs.add(dl.leftBack);
                    rightEncs.add(dl.rightFront);
                    rightEncs.add(dl.rightBack);
                } else if (md.localizer instanceof ThreeDeadWheelLocalizer) {
                    ThreeDeadWheelLocalizer dl = (ThreeDeadWheelLocalizer) md.localizer;
                    parEncs.add(dl.par0);
                    parEncs.add(dl.par1);
                    perpEncs.add(dl.perp);
                } else if (md.localizer instanceof TwoDeadWheelLocalizer) {
                    TwoDeadWheelLocalizer dl = (TwoDeadWheelLocalizer) md.localizer;
                    parEncs.add(dl.par);
                    perpEncs.add(dl.perp);
                } else {
                    throw new RuntimeException("unknown localizer: " + md.localizer.getClass().getName());
                }

                return new DriveView(
                        DriveType.MECANUM,
                        MecanumDrive.PARAMS.inPerTick,
                        MecanumDrive.PARAMS.maxWheelVel,
                        MecanumDrive.PARAMS.minProfileAccel,
                        MecanumDrive.PARAMS.maxProfileAccel,
                        hardwareMap.getAll(LynxModule.class),
                        Arrays.asList(
                                md.leftFront,
                                md.leftBack
                        ),
                        Arrays.asList(
                                md.rightFront,
                                md.rightBack
                        ),
                        leftEncs,
                        rightEncs,
                        parEncs,
                        perpEncs,
                        md.lazyImu,
                        md.voltageSensor,
                        () -> new MotorFeedforward(MecanumDrive.PARAMS.kS,
                                MecanumDrive.PARAMS.kV / MecanumDrive.PARAMS.inPerTick,
                                MecanumDrive.PARAMS.kA / MecanumDrive.PARAMS.inPerTick)
                );
            };
        } else {
            throw new RuntimeException();
        }

        manager.register(metaForClass(AngularRampLogger.class), new AngularRampLogger(dvf));
        manager.register(metaForClass(ForwardPushTest.class), new ForwardPushTest(dvf));
        manager.register(metaForClass(ForwardRampLogger.class), new ForwardRampLogger(dvf));
        manager.register(metaForClass(LateralPushTest.class), new LateralPushTest(dvf));
        manager.register(metaForClass(LateralRampLogger.class), new LateralRampLogger(dvf));
        manager.register(metaForClass(ManualFeedforwardTuner.class), new ManualFeedforwardTuner(dvf));
        manager.register(metaForClass(MecanumMotorDirectionDebugger.class), new MecanumMotorDirectionDebugger(dvf));
        manager.register(metaForClass(DeadWheelDirectionDebugger.class), new DeadWheelDirectionDebugger(dvf));

        manager.register(metaForClass(ManualFeedbackTuner.class), ManualFeedbackTuner.class);
        manager.register(metaForClass(SplineTest.class), SplineTest.class);
        manager.register(metaForClass(LocalizationTest.class), LocalizationTest.class);

        FtcDashboard.getInstance().withConfigRoot(configRoot -> {
            for (Class<?> c : Arrays.asList(
                    AngularRampLogger.class,
                    ForwardRampLogger.class,
                    LateralRampLogger.class,
                    ManualFeedforwardTuner.class,
                    MecanumMotorDirectionDebugger.class,
                    ManualFeedbackTuner.class
            )) {
                configRoot.putVariable(c.getSimpleName(), ReflectionConfig.createVariableFromClass(c));
            }
        });
    }
}

File: TeamCode/src/main/main/java/org/firstinspires/ftc/teamcode/drivetrain/tuning/TuningOpModes.kt
package org.firstinspires.ftc.teamcode.drivetrain.tuning

import com.acmerobotics.dashboard.CustomVariableConsumer
import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.config.reflection.ReflectionConfig
import com.acmerobotics.dashboard.config.variable.CustomVariable
import com.acmerobotics.roadrunner.MotorFeedforward
import com.acmerobotics.roadrunner.Pose2d
import com.acmerobotics.roadrunner.ftc.AngularRampLogger
import com.acmerobotics.roadrunner.ftc.DeadWheelDirectionDebugger
import com.acmerobotics.roadrunner.ftc.DriveType
import com.acmerobotics.roadrunner.ftc.DriveType.MECANUM
import com.acmerobotics.roadrunner.ftc.DriveView
import com.acmerobotics.roadrunner.ftc.DriveViewFactory
import com.acmerobotics.roadrunner.ftc.Encoder
import com.acmerobotics.roadrunner.ftc.FeedforwardFactory
import com.acmerobotics.roadrunner.ftc.ForwardPushTest
import com.acmerobotics.roadrunner.ftc.ForwardRampLogger
import com.acmerobotics.roadrunner.ftc.LateralPushTest
import com.acmerobotics.roadrunner.ftc.LateralRampLogger
import com.acmerobotics.roadrunner.ftc.ManualFeedforwardTuner
import com.acmerobotics.roadrunner.ftc.MecanumMotorDirectionDebugger
import com.qualcomm.hardware.lynx.LynxModule
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.eventloop.opmode.OpModeManager
import com.qualcomm.robotcore.eventloop.opmode.OpModeRegistrar
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.robotcore.internal.opmode.OpModeMeta
import org.firstinspires.ftc.robotcore.internal.opmode.OpModeMeta.Flavor.TELEOP
import org.firstinspires.ftc.teamcode.drivetrain.MecanumDrive
import org.firstinspires.ftc.teamcode.drivetrain.MecanumDrive.DriveLocalizer
import org.firstinspires.ftc.teamcode.drivetrain.ThreeDeadWheelLocalizer

object TuningOpModes {
    const val GROUP = "quickstart"
    const val DISABLED = false

    private fun metaForClass(cls: Class<out OpMode>): OpModeMeta {
        return OpModeMeta.Builder().setName(cls.getSimpleName()).setGroup(GROUP).setFlavor(TELEOP)
            .build()
    }

    @OpModeRegistrar
    fun register(manager: OpModeManager) {
        if (DISABLED) return

        val dvf = object : DriveViewFactory {
            override fun make(hardwareMap: HardwareMap): DriveView {

                val md: MecanumDrive = MecanumDrive(hardwareMap, Pose2d(0.0, 0.0, 0.0))
                val leftEncs: MutableList<Encoder> = mutableListOf()
                val rightEncs: MutableList<Encoder> = mutableListOf()
                val parEncs: MutableList<Encoder> = mutableListOf()
                val perpEncs: MutableList<Encoder> = mutableListOf()
                if (md.localizer is DriveLocalizer) {
                    val dl: DriveLocalizer = md.localizer
                    leftEncs.add(dl.leftFront)
                    leftEncs.add(dl.leftBack)
                    rightEncs.add(dl.rightFront)
                    rightEncs.add(dl.rightBack)
                } else if (md.localizer is ThreeDeadWheelLocalizer) {
                    val dl: ThreeDeadWheelLocalizer = md.localizer
                    parEncs.add(dl.par0)
                    parEncs.add(dl.par1)
                    perpEncs.add(dl.perp)
                } else {
                    throw java.lang.RuntimeException(
                        "unknown localizer: " + md.localizer.javaClass.getName()
                    )
                }
                return DriveView(MECANUM,
                    MecanumDrive.PARAMS.inPerTick,
                    MecanumDrive.PARAMS.maxWheelVel,
                    MecanumDrive.PARAMS.minProfileAccel,
                    MecanumDrive.PARAMS.maxProfileAccel,
                    hardwareMap.getAll<LynxModule>(LynxModule::class.java),
                    listOf(
                        md.leftFront, md.leftBack
                    ),
                    listOf(
                        md.rightFront, md.rightBack
                    ),
                    leftEncs,
                    rightEncs,
                    parEncs,
                    perpEncs,
                    md.lazyImu,
                    md.voltageSensor,
                    FeedforwardFactory {
                        MotorFeedforward(
                            MecanumDrive.PARAMS.kS,
                            MecanumDrive.PARAMS.kV / MecanumDrive.PARAMS.inPerTick,
                            MecanumDrive.PARAMS.kA / MecanumDrive.PARAMS.inPerTick
                        )
                    })
            }
        }

        manager.register(metaForClass(AngularRampLogger::class.java), AngularRampLogger(dvf) )
        manager.register(metaForClass(ForwardPushTest::class.java ), ForwardPushTest(dvf) )
        manager.register(metaForClass(ForwardRampLogger::class.java ), ForwardRampLogger(dvf) )
        manager.register(metaForClass(LateralPushTest::class.java ), LateralPushTest(dvf) )
        manager.register(metaForClass(LateralRampLogger::class.java ), LateralRampLogger(dvf) )
        manager.register(metaForClass(ManualFeedforwardTuner::class.java ), ManualFeedforwardTuner(dvf) )
        manager.register(metaForClass(MecanumMotorDirectionDebugger::class.java ), MecanumMotorDirectionDebugger(dvf) )
        manager.register(metaForClass(DeadWheelDirectionDebugger::class.java ), DeadWheelDirectionDebugger(dvf) )
        manager.register(metaForClass(ManualFeedbackTuner::class.java ), ManualFeedbackTuner::class.java )
        manager.register(metaForClass(SplineTest::class.java ), SplineTest::class.java )
        manager.register(metaForClass(LocalizationTest::class.java ), LocalizationTest::class.java )

        FtcDashboard.getInstance()
            .withConfigRoot(CustomVariableConsumer { configRoot: CustomVariable ->
                listOf<Class<out Any>>(
                    AngularRampLogger::class.java,
                    ForwardRampLogger::class.java,
                    LateralRampLogger::class.java,
                    ManualFeedforwardTuner::class.java,
                    MecanumMotorDirectionDebugger::class.java,
                    ManualFeedbackTuner::class.java
                ).forEach {
                    configRoot.putVariable(
                        it.getSimpleName(),
                        ReflectionConfig.createVariableFromClass(it)
                    )
                }
            })
    }
}
